File: app/src/main/java/com/example/customprogressbar/remainingQuiteTime/QuiteTimeTimer.java
package com.example.customprogressbar.remainingQuiteTime;

import android.os.CountDownTimer;

import java.util.LinkedList;
import java.util.List;

class QuiteTimeTimer {

    private static final int millisInSecond = 1000;

    private CountDownTimer timer;
    private List<QuiteTimeTimerListener> timerListeners = new LinkedList<>();
    private boolean removeAll = false;

     QuiteTimeTimer() {
         timer = new CountDownTimer(Long.MAX_VALUE, millisInSecond) {
             @Override
             public void onTick(long l) {
                 Integer removePosition = null;
                 for (QuiteTimeTimerListener timerListener : timerListeners) {
                     if(removeAll) {
                         deleteAllSubscribers();
                         removeAll = false;
                         break;
                     }
                     if(removePosition == null) {
                         removePosition = timerListener.onTick();
                     } else {
                         timerListener.onTick();
                     }
                 }



                 if(removePosition != null) {
                     //unsubscribed, so that the timer can stop ticking if no one is listening
                     unsubscribeFromTimer(timerListeners.get(removePosition));
                 }

             }

             @Override
             public void onFinish() {
                 //ignore
             }
         };
     }

     void removeAllSubscribers() {
         removeAll = true;
     }


     void subscribeToTimer(QuiteTimeTimerListener quiteTimeTimerListener) {
         if(timerListeners.isEmpty()) {
             timer.start();
         }
         timerListeners.add(quiteTimeTimerListener);
     }

    void unsubscribeFromTimer(QuiteTimeTimerListener quiteTimeTimerListener) {
         timerListeners.remove(quiteTimeTimerListener);
         if(timerListeners.isEmpty()) {
             timer.cancel();
         }
     }

     private void deleteAllSubscribers() {
//         timer.cancel();
         timerListeners = new LinkedList<>();
     }



}

File: app/src/main/java/com/example/customprogressbar/remainingQuiteTime/QuiteTimeTimer.kt
package com.example.customprogressbar.remainingQuiteTime

import android.os.CountDownTimer
import java.util.LinkedList

interface QuiteTimeTimerListener {
    fun onTick()
    fun onFinished(index: Int)
    fun isFinished(): Boolean
    fun isStopped(): Boolean
}

object QuiteTimeTimer {

    private const val millisInSecond = 1000

    private val timer: CountDownTimer
    private val timerListeners = LinkedList<QuiteTimeTimerListener>()


    init {
        timer = object : CountDownTimer(java.lang.Long.MAX_VALUE, millisInSecond.toLong()) {
            override fun onTick(l: Long) {
                var finishedTimerListener: QuiteTimeTimerListener? = null
                var stoppedTimerListener: QuiteTimeTimerListener? = null
                for (timerListener in timerListeners) {
                    timerListener.onTick()
                    if(timerListener.isFinished()) {
                        finishedTimerListener = timerListener
                    }
                    if(timerListener.isStopped()) {
                        stoppedTimerListener = timerListener
                    }
                }
                if(finishedTimerListener != null) {
                    val index = timerListeners.indexOf(finishedTimerListener)

                    unsubscribeFromTimer(finishedTimerListener)
                    finishedTimerListener.onFinished(index)
                }

                if(stoppedTimerListener != null) {
                    val index = timerListeners.indexOf(stoppedTimerListener)

                    unsubscribeFromTimer(stoppedTimerListener)
                    stoppedTimerListener.onFinished(index)
                }
            }

            override fun onFinish() {
                //ignore
            }
        }
    }

    fun resetTimer() {
        for(timerListener in timerListeners) {
            unsubscribeFromTimer(timerListener)
        }
    }

    fun subscribeToTimer(quiteTimeTimerListener: QuiteTimeTimerListener) {
        timerListeners.add(quiteTimeTimerListener)
        if (timerListeners.size == 1) {
            timer.start()
        }
    }

    fun unsubscribeFromTimer(quiteTimeTimerListener: QuiteTimeTimerListener) {
        if (timerListeners.size == 1) {
            timer.cancel()
        }
        timerListeners.remove(quiteTimeTimerListener)
    }
}